RealmSwift/SwiftUI.swift

```diff
@@ -1450,59 +1449,67 @@ private class ObservableAsyncOpenStorage: ObservableObject {
-        var open = false
-        if let partitionValue = partitionValue {
-            let bsonValue = AnyBSON(partitionValue: partitionValue)
-            if self.partitionValue != bsonValue {
-                self.partitionValue = bsonValue
-                open = true
-            }
-        }
+    init(asyncOpenKind: AsyncOpenKind, app: App, configuration: Realm.Configuration?, partitionValue: AnyBSON?) {
+        self.asyncOpenKind = asyncOpenKind
+        self.app = app
+        self.configuration = configuration
+        self.partitionValue = partitionValue
 
-        // We don't want to use the `defaultConfiguration` from the environment, we only want to use this environment value in @AsyncOpen if is not the default one
-        if configuration != .defaultConfiguration, self.configuration != configuration {
-            if let partitionValue = configuration.syncConfiguration?.partitionValue {
-                self.partitionValue = partitionValue
-            }
-            self.configuration = configuration
-            open = true
-        }
-        if open {
-            self.asyncOpen()
+        // Initialising the state value depending on the user status, before first rendering.
+        if let user = app.currentUser {
+            appState = .loggedIn(user)
+            asyncOpenState = .connecting
+        } else {
+            asyncOpenState = .waitingForUser
         }
     }
 
-    private func asyncOpen() {
-        if case let .loggedIn(user) = appState {
+    var setupHasRun = false
+    func setup() {
+        guard !setupHasRun else { return }
+        initAsyncOpen()
+        setupHasRun = true
+    }
+
+    private func initAsyncOpen() {
+        if case .loggedIn(let user) = appState {
+            // we only open the realm on initialisation if there is a user logged.
             asyncOpenForUser(user)
         }
+
+        // we observe the changes in the app state to check for user changes,
+        // we store an internal state, so we could react to those changes (user login, user change, logout).
+        app.objectWillChange.sink { app in
+            switch self.appState {
+            case .loggedIn(let user):
+                if let newUser = app.currentUser,
+                    user != newUser {
+                    self.appState = .loggedIn(newUser)
+                    self.asyncOpenState = .connecting
+                    self.asyncOpenForUser(user)
+                } else if app.currentUser == nil {
+                    self.asyncOpenState = .waitingForUser
+                    self.appState = .loggedOut
+                }
+            case .loggedOut:
+                if let user = app.currentUser {
+                    self.appState = .loggedIn(user)
+                    self.asyncOpenState = .connecting
+                    self.asyncOpenForUser(user)
+                }
+            }
+        }.store(in: &appCancellable)
```

Comment:
I think this is a retain cycle. `self` retains `appCancellable`, `appCancellable` retains the closure, and the closure retains `self`. This probably needs a `[weak self]` capture.

Reply:
Changed this to avoid the retain cycle